### PR TITLE
Generalizations

### PR DESCRIPTION
**Pull request by brossetti1**
## Description

- [X] adding an interface for the bitswap client
- [X] adding configuration for static values 
- [X] adding ctx to Get in order to allow users to wrap a session with ctx logic

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -37,17 +42,33 @@ type Session struct {
...
 
 // New initiates a bitswap retrieval session
-func New(h host.Host, peer peer.ID) *Session {
+func New(h host.Host, peer peer.ID, opts Options) *Session {
+	if opts.SessionTimeout == 0 {
+		opts.SessionTimeout = 10 * time.Second
+	}
+
+	if opts.TickerTimeout == 0 {
+		opts.TickerTimeout = 50 * time.Millisecond
```
Comment (willscott) [inline session.go]: have defaults as constants defined next to options struct

File: session.go [hunk 2]
```diff
@@ -66,9 +87,7 @@ func (s *Session) connect() {
 	sessionCtx, cncl := context.WithCancel(context.Background())
 	s.close = cncl
 	go s.writeLoop(sessionCtx)
-
-	// todo: configuratble timeout
-	ctx, cncl := context.WithDeadline(context.Background(), time.Now().Add(10*time.Second))
+	ctx, cncl := context.WithDeadline(context.Background(), time.Now().Add(s.stimeout))
```
Comment (willscott) [inline session.go]: there should be a way to run a bitswap session without the session timeout.

File: session.go [hunk 3]
```diff
@@ -37,17 +42,33 @@ type Session struct {
 
 	interestMtx sync.Mutex
 	interests   map[cid.Cid]func([]byte, error)
+	stimeout    time.Duration
+	ttimeout    time.Duration
+}
+
+type Options struct {
+	SessionTimeout time.Duration
+	TickerTimeout  time.Duration
```
Comment (willscott) [inline session.go]: maybe `WriteAggregationQuantum`?

**Comment by brossetti1** [comment]
@willscott hows this?

**Review by willscott — APPROVED** [review]

**Pull request merged**